pkg/node/client.go

```diff
@@ -0,0 +1,165 @@
+	//}
+	if len(bar.Spec.BucketRequestName) == 0 {
+		return nil, logErr(fmt.Errorf("bucketAccessRequest.Spec.BucketRequestName unset"))
+	}
+	return bar, nil
+}
+
+func (n *NodeClient) getBA(ctx context.Context, baName string) (*v1alpha1.BucketAccess, error) {
+	klog.Infof("getting bucketAccess %q", fmt.Sprintf("%s", baName))
+	ba, err := n.cosiClient.BucketAccesses().Get(ctx, baName, metav1.GetOptions{})
+	if err != nil {
+		return nil, logErr(getError("bucketAccess", fmt.Sprintf("%s", baName), err))
+	}
+	// TODO: need to enable validation after resolving status issue - Krish
+	//if ba == nil || !ba.Status.AccessGranted {
+	//	return nil, logErr(fmt.Errorf("BucketAccess is nil or BucketAccess does not grant access %q", fmt.Sprintf("%s", baName)))
+	//}
+	return ba, nil
+}
+
+func (n *NodeClient) getBR(ctx context.Context, brName, brNs string) (*v1alpha1.BucketRequest, error) {
+	klog.Infof("getting bucketRequest %q", brName)
+	br, err := n.cosiClient.BucketRequests(brNs).Get(ctx, brName, metav1.GetOptions{})
+	if err != nil {
+		return nil, logErr(getError("bucketRequest", fmt.Sprintf("%s/%s", brNs, brName), err))
+	}
+	// TODO: need to enable validation after resolving status issue - Krish
+	//if br == nil || !br.Status.BucketAvailable {
+	//	return nil, logErr(fmt.Errorf("BucketRequest is nil or BucketRequest is not available yet %q", fmt.Sprintf("%s/%s", brNs, brName)))
+	//}
+	return br, nil
+}
+
+func (n *NodeClient) getB(ctx context.Context, bName string) (*v1alpha1.Bucket, error) {
+	klog.Infof("getting bucket %q", bName)
+	// is BucketInstanceName the correct field, or should it be BucketClass
+	bkt, err := n.cosiClient.Buckets().Get(ctx, bName, metav1.GetOptions{})
+	if err != nil {
+		return nil, logErr(getError("bucket", bName, err))
+	}
+	// TODO: need to enable validation after resolving status issue - Krish
+	//if bkt == nil || !bkt.Status.BucketAvailable {
+	//	return nil, logErr(fmt.Errorf("Bucket is nil or Bucket is not available yet %q", fmt.Sprintf("%s", bName)))
+	//}
+	return bkt, nil
+}
+
+func (n *NodeClient) GetResources(ctx context.Context, barName, barNs string) (bkt *v1alpha1.Bucket, secret *v1.Secret, err error) {
+	bar, err := n.getBAR(ctx, barName, barNs)
+	if err != nil {
+		return
+	}
+
+	ba, err := n.getBA(ctx, bar.Spec.BucketAccessName)
+	if err != nil {
+		return
+	}
+
+	br, err := n.getBR(ctx, bar.Spec.BucketRequestName, barNs)
+	if err != nil {
+		return
+	}
+
+	bkt, err = n.getB(ctx, br.Spec.BucketInstanceName)
+	if err != nil {
+		return
+	}
+
+	secret, err = n.kubeClient.CoreV1().Secrets(barNs).Get(ctx, ba.Spec.MintedSecretName, metav1.GetOptions{})
+	if err != nil {
+		_ = logErr(getError("pod", fmt.Sprintf("%s/%s", barNs, ba.Spec.MintedSecretName), err))
+		return
+	}
+	return
+}
+
+func (n *NodeClient) getProtocol(bkt *v1alpha1.Bucket) (interface{}, error) {
```

Comment:
why is the protocol being returned as an interface?

Reply:
I am returning it as an interface because we just marshal the struct into a string and this way we don't need to worry about what the underlying protocol is (S3 vs. Azure vs. AWS)

Reply:
My (admitted limited) understanding is that it is reasonable to define func parms as empty interfaces (see golang _fmt_ pkg) but an anti-pattern to functionally return `inerface{}`. Maybe you can research this some?

Reply:
I could alternatively marshal the protocol inside of this function, and return a []byte directly, but I am not a big fan of that approach since then this function is responsible for multiple parts of publish process as opposed to just parsing the Bucket

Reply:
Ok. I consider my comment resolved but I recommend getting @wlan0 's opinion.